src/radix_tree.rs

```diff
@@ -0,0 +1,164 @@
+use crate::vwmap::NamespaceDescriptor;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub(crate) struct NamespaceDescriptorWithHash {
+    pub(crate) descriptor: NamespaceDescriptor,
+    pub(crate) hash_seed: u32,
+}
+
+impl NamespaceDescriptorWithHash {
+    pub(crate) fn new(descriptor: NamespaceDescriptor, hash_seed: u32) -> Self {
+        Self {
+            descriptor,
+            hash_seed
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct RadixTreeNode {
+    children: Vec<Option<RadixTreeNode>>,
+    value: Option<NamespaceDescriptorWithHash>,
+}
+
+impl Default for RadixTreeNode {
+    fn default() -> Self {
+        Self {
+            children: vec![None; 256],
+            value: None,
+        }
+    }
+}
+
+impl RadixTreeNode {
+    fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct RadixTree {
+    root: RadixTreeNode,
+}
+
+impl RadixTree {
+    fn new() -> Self {
+        RadixTree {
+            root: RadixTreeNode::new(),
+        }
+    }
+
+    pub(crate)fn insert(&mut self, key: &[u8], value: NamespaceDescriptorWithHash) {
+        let mut node = &mut self.root;
+
+        for &byte in key {
+            let child = &mut node.children[byte as usize];
+            node = child.get_or_insert_with(|| RadixTreeNode::new());
+        }
+
+        node.value = Some(value);
+    }
+
+    pub(crate) fn get(&self, key: &[u8]) -> Option<&NamespaceDescriptorWithHash> {
+        let mut node = &self.root;
+
+        for &byte in key {
+            let maybe_child = &node.children[byte as usize];
+            if let Some(child) = maybe_child {
+                node = child;
+            } else {
+                return None.as_ref();
+            }
+        }
+
+        node.value.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::vwmap::{NamespaceFormat, NamespaceType};
+    use super::*;
+
+    #[test]
+    fn test_insert_and_get() {
+        let mut tree = RadixTree::new();
+
+        let namespace_descriptor_with_hash_1 = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 0,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
```

Comment:
Perhaps a test for non-categorical is in order to for completeness' sake

Reply:
Changing these doesn't provide any difference from the test perspective as we test for equality.